xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,379 @@
+	// pendingState is the state of the pending balancer, and is used to update
+	// the Client Conn in the scenario the pending balancer gets swapped to.
+	pendingState     balancer.State
+	currentLBIsReady bool
+
+	// currentMu protects operations on the current balancer.
+	currentMu sync.Mutex
+
+	// closed is fired when this balancer is closed.
+	closed *grpcsync.Event
+}
+
+func (gsb *gracefulSwitchBalancer) updateState(bal balancer.Balancer, state balancer.State) {
+	// Hold the mutex for this entire call to ensure it cannot occur
+	// concurrently with other updateState() calls. This causes updates to
+	// pendingState and calls to cc.UpdateState to happen atomically.
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	if bal == gsb.balancerPending {
+		// Cache the pending state and picker if you don't need to send it at
+		// this instant (i.e. the LB policy is connecting). This balancer can
+		// send it later on the current LB exiting READY.
+		gsb.pendingState = state
+		// If the current balancer is currently in a state other than READY, the
+		// new policy can be swapped into place immediately. This is because
+		// there is no reason to gracefully switch from and keep using the old
+		// policy as the ClientConn is not connected to any backends.
+		if state.ConnectivityState != connectivity.Connecting || !gsb.currentLBIsReady {
+			gsb.swap()
+		}
+		return
+	}
+	gsb.currentLBIsReady = state.ConnectivityState == connectivity.Ready
+	// In the case that the current balancer exits READY, and there is a pending
+	// balancer, you can forward the pending balancers cached State up to
+	// ClientConn and swap the pending into the current. This is because there
+	// is no reason to gracefully switch from and keep using the old policy as
+	// the ClientConn is not connected to any backends.
+	if !gsb.currentLBIsReady && gsb.balancerPending != nil {
+		gsb.swap()
+		return
+	}
+	// Even if there is a pending balancer waiting to be gracefully switched to,
+	// continue to forward current balancer updates to the Client Conn. Ignoring
+	// state + picker from the current would cause undefined behavior/cause the
+	// system to behave incorrectly from the current LB policies perspective.
+	// Also, the current LB is still being used by grpc to choose SubConns per
+	// RPC, and thus should use the most updated form of the current balancer.
+	gsb.cc.UpdateState(state)
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	currBalToClose := gsb.balancerCurrent
+	go func() {
+		gsb.currentMu.Lock()
+		defer gsb.currentMu.Unlock()
+		currBalToClose.Close()
+		gsb.mu.Lock()
+		defer gsb.mu.Unlock()
+		for sc, bal := range gsb.scToSubBalancer {
+			if bal == gsb.balancerCurrent {
+				gsb.cc.RemoveSubConn(sc)
+				delete(gsb.scToSubBalancer, sc)
+			}
+		}
```

Comment:
It's not too important, but I think you want to also do this synchronously.  Also please move the goroutine to the end of the function so it's more clear what all is happening synchronously and what is happening asynchronously.